### ISSUE TITLE
Use inspection on __init__ for from_params

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#1158]
I just ran into a bug where I had changed the default value in `__init__` for a particular argument, but forgot to change it in `from_params`.  It's unfortunate we need to repeat ourselves in `from_params` so much, and I wonder if there's a way we could make this simpler using `inspect`.  Like, get the type annotation of the argument, use that to know how to do `params.pop`, pass the default value from the inspection along to the `params.pop`, etc.  I imagine you could just have a default implementation on `Registrable` that just does inspection on the constructor of the object, and have it just work for almost all classes.  For particularly complex ones, like `Vocabulary` or `Embedding`, you'd need to override it, but for standard things, you should be able to just use inspection.  Thoughts?

**Comment by matt-gardner** [comment]
That actually looks standard to me except for the `vocab` argument. For the rest, you can just find the annotated type and call `type.from_params(params.pop("..."))`. You just need `Model` to override the standard method a little, and then all models inherit the version that uses a vocab.

**Comment by janLuke** [comment]
1. @matt-gardner I'm very new to the framework, so I'm not sure I see all the potential problems here but... do you mean something like this? (This is just an untested prototype)
```
def from_params(cls, params: Params):
    kwargs = dict()
    init_args = dict(inspect.signature(cls.__init__).parameters)
    init_args.pop('self')
    for arg in init_args.values():
        arg_type = arg.annotation
        if arg_type != Parameter.empty and hasattr(arg_type, 'from_params'):
            default = arg.default if arg.default != Parameter.empty else None
            kwargs[arg.name] = arg_type.from_params(params.pop(arg.name, default))
        else:
            if arg_type == Parameter.empty:
                logger.warning('No type annotation for argument %s', arg.name)
            kwargs[arg.name] = params.pop(arg.name, arg.default)
    params.assert_empty(cls.__name__)
    return cls(**kwargs)
```
Notice that this assumes that Union arguments are not loadable from params. 

Relying on type annotations doesn't convince me, anyway. Maybe giving a look to https://github.com/IDSIA/sacred may help.

2. Could you explain why there's the need for ``from_params()`` functions that accept arguments other than ``params`` (like in Vocabulary)?

**Comment by joelgrus** [comment]
so I actually started on this and then got distracted with other things, but here's a WIP:

https://github.com/allenai/allennlp/pull/1191

it puts `from_params` in `Registrable`, and then you only have to override it when you have a base class with unusual behavior

**Comment by DeNeutoy** [comment]
fixed by #1191
